File: api/app/oauth/registration.py
from __future__ import annotations

from dataclasses import dataclass
from typing import Any

from fastapi.responses import JSONResponse
from pydantic import BaseModel, Field

from ..config import Settings
from .common import (
    SUPPORTED_GRANT_TYPES,
    SUPPORTED_RESPONSE_TYPES,
    SUPPORTED_TOKEN_ENDPOINT_AUTH_METHODS,
    oauth_error_response,
    require_oauth_enabled,
)
from .repository import create_oauth_client
from .service import build_oauth_client_id, issue_oauth_client_secret, oauth_secret_hash


class OAuthClientRegistrationRequest(BaseModel):
    redirect_uris: list[str] = Field(default_factory=list)
    client_name: str = Field(default="Engram MCP Client", min_length=1, max_length=200)
    grant_types: list[str] = Field(default_factory=lambda: ["authorization_code"])
    response_types: list[str] = Field(default_factory=lambda: ["code"])
    token_endpoint_auth_method: str = "none"


@dataclass(frozen=True)
class _ValidatedRegistrationRequest:
    redirect_uris: list[str]
    grant_types: list[str]
    response_types: list[str]
    token_endpoint_auth_method: str


def _dedup_string_list(values: list[str]) -> list[str]:
    normalized: list[str] = []
    seen: set[str] = set()
    for item in values:
        value = (item or "").strip()
        if not value:
            continue
        if value in seen:
            continue
        seen.add(value)
        normalized.append(value)
    return normalized


def _sanitize_redirect_uris(values: list[str]) -> list[str]:
    return _dedup_string_list(values)


def _normalize_metadata_list(values: list[str], fallback: set[str]) -> list[str]:
    normalized = _dedup_string_list(values)
    if not normalized:
        return sorted(fallback)
    return normalized


def _authorize_registration_request(
    *,
    settings: Settings,
    session_user: dict[str, Any] | None,
) -> JSONResponse | None:
    if not settings.oauth_require_protected_registration:
        return None
    if not session_user:
        return oauth_error_response(
            error="invalid_client",
            description="Authenticated admin session required for dynamic registration.",
            status_code=401,
        )
    if (session_user.get("role") or "").strip().lower() != "admin":
        return oauth_error_response(
            error="insufficient_privilege",
            description="Admin role is required for dynamic registration.",
            status_code=403,
        )
    return None


def _validate_registration_payload(
    payload: OAuthClientRegistrationRequest,
) -> tuple[_ValidatedRegistrationRequest | None, JSONResponse | None]:
    redirect_uris = _sanitize_redirect_uris(payload.redirect_uris)
    if not redirect_uris:
        return None, oauth_error_response(
            error="invalid_redirect_uri",
            description="At least one redirect_uri is required.",
        )

    token_endpoint_auth_method = payload.token_endpoint_auth_method.strip() or "none"
    if token_endpoint_auth_method not in SUPPORTED_TOKEN_ENDPOINT_AUTH_METHODS:
        return None, oauth_error_response(
            error="invalid_client_metadata",
            description="Unsupported token_endpoint_auth_method.",
        )

    grant_types = _normalize_metadata_list(payload.grant_types, SUPPORTED_GRANT_TYPES)
    if "authorization_code" not in set(grant_types):
        return None, oauth_error_response(
            error="invalid_client_metadata",
            description="authorization_code grant type is required.",
        )

    response_types = _normalize_metadata_list(payload.response_types, SUPPORTED_RESPONSE_TYPES)
    if "code" not in set(response_types):
        return None, oauth_error_response(
            error="invalid_client_metadata",
            description="code response type is required.",
        )
    return (
        _ValidatedRegistrationRequest(
            redirect_uris=redirect_uris,
            grant_types=grant_types,
            response_types=response_types,
            token_endpoint_auth_method=token_endpoint_auth_method,
        ),
        None,
    )


def _issue_confidential_client_secret(
    *,
    settings: Settings,
    client_id: str,
    token_endpoint_auth_method: str,
) -> tuple[str | None, str | None]:
    if token_endpoint_auth_method != "client_secret_post":
        return None, None
    client_secret = issue_oauth_client_secret()
    client_secret_hash = oauth_secret_hash(
        identifier=client_id,
        secret=client_secret,
        pepper=settings.oauth_client_secret_pepper,
    )
    return client_secret, client_secret_hash


def _handle_oauth_register(
    *,
    settings: Settings,
    payload: OAuthClientRegistrationRequest,
    session_user: dict[str, Any] | None,
) -> JSONResponse:
    require_oauth_enabled(settings)
    auth_error = _authorize_registration_request(settings=settings, session_user=session_user)
    if auth_error is not None:
        return auth_error

    validated_payload, validation_error = _validate_registration_payload(payload)
    if validation_error is not None:
        return validation_error
    if validated_payload is None:  # pragma: no cover - defensive
        return oauth_error_response(
            error="invalid_client_metadata",
            description="Unable to validate registration payload.",
        )

    client_id = build_oauth_client_id()
    client_secret, client_secret_hash = _issue_confidential_client_secret(
        settings=settings,
        client_id=client_id,
        token_endpoint_auth_method=validated_payload.token_endpoint_auth_method,
    )

    created = create_oauth_client(
        client_id=client_id,
        client_name=payload.client_name.strip(),
        redirect_uris=validated_payload.redirect_uris,
        grant_types=validated_payload.grant_types,
        response_types=validated_payload.response_types,
        token_endpoint_auth_method=validated_payload.token_endpoint_auth_method,
        client_secret_hash=client_secret_hash,
        metadata_json=payload.model_dump(mode="python"),
    )
    issued_at = int(created.created_at.timestamp())
    response_payload: dict[str, Any] = {
        "client_id": created.client_id,
        "client_name": created.client_name,
        "redirect_uris": created.redirect_uris,
        "grant_types": created.grant_types,
        "response_types": created.response_types,
        "token_endpoint_auth_method": created.token_endpoint_auth_method,
        "client_id_issued_at": issued_at,
    }
    if client_secret is not None:
        response_payload["client_secret"] = client_secret
        response_payload["client_secret_expires_at"] = 0
    return JSONResponse(status_code=201, content=response_payload)

File: internal/oauth/registration.go
package oauth

import (
	"context"
	"errors"
	"sort"
	"strings"

	"engram/internal/config"
	"engram/internal/models"
	"engram/internal/repository"
)

var (
	supportedGrantTypes              = map[string]struct{}{"authorization_code": {}}
	supportedResponseTypes           = map[string]struct{}{"code": {}}
	supportedTokenEndpointAuthMethod = map[string]struct{}{"none": {}, "client_secret_post": {}}
)

// SessionUser captures authorization context for protected registration.
type SessionUser struct {
	UserID   string
	Username string
	Role     string
}

// RegistrationRequest captures OAuth dynamic client registration metadata.
type RegistrationRequest struct {
	RedirectURIs            []string
	ClientName              string
	GrantTypes              []string
	ResponseTypes           []string
	TokenEndpointAuthMethod string
}

// RegistrationResponse captures successful OAuth dynamic registration output.
type RegistrationResponse struct {
	ClientID                string   `json:"client_id"`
	ClientName              string   `json:"client_name"`
	RedirectURIs            []string `json:"redirect_uris"`
	GrantTypes              []string `json:"grant_types"`
	ResponseTypes           []string `json:"response_types"`
	TokenEndpointAuthMethod string   `json:"token_endpoint_auth_method"`
	ClientIDIssuedAt        int64    `json:"client_id_issued_at"`
	ClientSecret            *string  `json:"client_secret,omitempty"`
	ClientSecretExpiresAt   *int     `json:"client_secret_expires_at,omitempty"`
}

// RegistrationError captures oauth-style registration errors.
type RegistrationError struct {
	ErrorCode   string
	Description string
	StatusCode  int
}

func (err RegistrationError) Error() string {
	return err.Description
}

// RegistrationService handles OAuth dynamic client registration.
type RegistrationService struct {
	db   repository.Queryer
	deps registrationDeps
}

type registrationDeps struct {
	buildClientID     func() string
	issueClientSecret func() (string, error)
	oauthSecretHash   func(input SecretHashInput) string
	createOAuthClient func(ctx context.Context, db repository.Queryer, input repository.OAuthClientCreateInput) (*models.OAuthClientRecord, error)
}

func defaultRegistrationDeps() registrationDeps {
	return registrationDeps{
		buildClientID:     BuildOAuthClientID,
		issueClientSecret: IssueOAuthClientSecret,
		oauthSecretHash:   OAuthSecretHash,
		createOAuthClient: repository.CreateOAuthClient,
	}
}

// NewRegistrationService builds a registration service with repository defaults.
func NewRegistrationService(db repository.Queryer) *RegistrationService {
	return &RegistrationService{db: db, deps: defaultRegistrationDeps()}
}

// HandleRegister validates and persists OAuth dynamic client registration metadata.
func (service *RegistrationService) HandleRegister(
	ctx context.Context,
	settings config.Settings,
	payload RegistrationRequest,
	sessionUser *SessionUser,
) (RegistrationResponse, error) {
	if !settings.OAuthEnabled {
		return RegistrationResponse{}, RegistrationError{
			ErrorCode:   "oauth_disabled",
			Description: "OAuth endpoints disabled",
			StatusCode:  404,
		}
	}
	if err := authorizeRegistrationRequest(settings, sessionUser); err != nil {
		return RegistrationResponse{}, err
	}
	validatedPayload, err := validateRegistrationPayload(payload)
	if err != nil {
		return RegistrationResponse{}, err
	}
	createdClientID := service.deps.buildClientID()
	clientSecret, clientSecretHash, err := issueConfidentialClientSecret(
		service.deps,
		settings,
		createdClientID,
		validatedPayload.TokenEndpointAuthMethod,
	)
	if err != nil {
		return RegistrationResponse{}, err
	}
	created, err := service.deps.createOAuthClient(
		ctx,
		service.db,
		repository.OAuthClientCreateInput{
			ClientID:                createdClientID,
			ClientName:              strings.TrimSpace(payload.ClientName),
			RedirectURIs:            validatedPayload.RedirectURIs,
			GrantTypes:              validatedPayload.GrantTypes,
			ResponseTypes:           validatedPayload.ResponseTypes,
			TokenEndpointAuthMethod: validatedPayload.TokenEndpointAuthMethod,
			ClientSecretHash:        clientSecretHash,
			MetadataJSON:            payloadToMetadata(payload),
		},
	)
	if err != nil {
		return RegistrationResponse{}, err
	}
	response := RegistrationResponse{
		ClientID:                created.ClientID,
		ClientName:              created.ClientName,
		RedirectURIs:            created.RedirectURIs,
		GrantTypes:              created.GrantTypes,
		ResponseTypes:           created.ResponseTypes,
		TokenEndpointAuthMethod: created.TokenEndpointAuthMethod,
		ClientIDIssuedAt:        created.CreatedAt.Unix(),
		ClientSecret:            clientSecret,
	}
	if clientSecret != nil {
		expiresAt := 0
		response.ClientSecretExpiresAt = &expiresAt
	}
	return response, nil
}

type validatedRegistrationRequest struct {
	RedirectURIs            []string
	GrantTypes              []string
	ResponseTypes           []string
	TokenEndpointAuthMethod string
}

// DedupStringList trims and deduplicates string items while preserving order.
func DedupStringList(values []string) []string {
	normalized := make([]string, 0, len(values))
	seen := make(map[string]struct{}, len(values))
	for _, item := range values {
		value := strings.TrimSpace(item)
		if value == "" {
			continue
		}
		if _, exists := seen[value]; exists {
			continue
		}
		seen[value] = struct{}{}
		normalized = append(normalized, value)
	}
	return normalized
}

func authorizeRegistrationRequest(settings config.Settings, sessionUser *SessionUser) error {
	if !settings.OAuthRequireProtectedRegistration {
		return nil
	}
	if sessionUser == nil {
		return RegistrationError{
			ErrorCode:   "invalid_client",
			Description: "Authenticated admin session required for dynamic registration.",
			StatusCode:  401,
		}
	}
	if strings.ToLower(strings.TrimSpace(sessionUser.Role)) != "admin" {
		return RegistrationError{
			ErrorCode:   "insufficient_privilege",
			Description: "Admin role is required for dynamic registration.",
			StatusCode:  403,
		}
	}
	return nil
}

func validateRegistrationPayload(payload RegistrationRequest) (validatedRegistrationRequest, error) {
	redirectURIs := DedupStringList(payload.RedirectURIs)
	if len(redirectURIs) == 0 {
		return validatedRegistrationRequest{}, RegistrationError{
			ErrorCode:   "invalid_redirect_uri",
			Description: "At least one redirect_uri is required.",
			StatusCode:  400,
		}
	}
	tokenEndpointAuthMethod := strings.TrimSpace(payload.TokenEndpointAuthMethod)
	if tokenEndpointAuthMethod == "" {
		tokenEndpointAuthMethod = "none"
	}
	if _, ok := supportedTokenEndpointAuthMethod[tokenEndpointAuthMethod]; !ok {
		return validatedRegistrationRequest{}, RegistrationError{
			ErrorCode:   "invalid_client_metadata",
			Description: "Unsupported token_endpoint_auth_method.",
			StatusCode:  400,
		}
	}
	grantTypes := normalizeMetadataList(payload.GrantTypes, supportedGrantTypes)
	if !containsValue(grantTypes, "authorization_code") {
		return validatedRegistrationRequest{}, RegistrationError{
			ErrorCode:   "invalid_client_metadata",
			Description: "authorization_code grant type is required.",
			StatusCode:  400,
		}
	}
	responseTypes := normalizeMetadataList(payload.ResponseTypes, supportedResponseTypes)
	if !containsValue(responseTypes, "code") {
		return validatedRegistrationRequest{}, RegistrationError{
			ErrorCode:   "invalid_client_metadata",
			Description: "code response type is required.",
			StatusCode:  400,
		}
	}
	return validatedRegistrationRequest{
		RedirectURIs:            redirectURIs,
		GrantTypes:              grantTypes,
		ResponseTypes:           responseTypes,
		TokenEndpointAuthMethod: tokenEndpointAuthMethod,
	}, nil
}

func normalizeMetadataList(values []string, fallback map[string]struct{}) []string {
	normalized := DedupStringList(values)
	if len(normalized) > 0 {
		return normalized
	}
	fallbackValues := make([]string, 0, len(fallback))
	for value := range fallback {
		fallbackValues = append(fallbackValues, value)
	}
	sort.Strings(fallbackValues)
	return fallbackValues
}

func containsValue(values []string, expected string) bool {
	for _, value := range values {
		if value == expected {
			return true
		}
	}
	return false
}

func issueConfidentialClientSecret(
	deps registrationDeps,
	settings config.Settings,
	clientID string,
	tokenEndpointAuthMethod string,
) (*string, *string, error) {
	if tokenEndpointAuthMethod != "client_secret_post" {
		return nil, nil, nil
	}
	clientSecret, err := deps.issueClientSecret()
	if err != nil {
		return nil, nil, err
	}
	clientSecretHash := deps.oauthSecretHash(SecretHashInput{
		Identifier: clientID,
		Secret:     clientSecret,
		Pepper:     settings.OAuthClientSecretPepper,
	})
	return &clientSecret, &clientSecretHash, nil
}

func payloadToMetadata(payload RegistrationRequest) map[string]any {
	return map[string]any{
		"redirect_uris":              payload.RedirectURIs,
		"client_name":                payload.ClientName,
		"grant_types":                payload.GrantTypes,
		"response_types":             payload.ResponseTypes,
		"token_endpoint_auth_method": payload.TokenEndpointAuthMethod,
	}
}

func asRegistrationError(err error) (RegistrationError, bool) {
	registrationErr := RegistrationError{}
	if errors.As(err, &registrationErr) {
		return registrationErr, true
	}
	return RegistrationError{}, false
}
